src/cascadia/QueryExtension/ExtensionPalette.cpp

```diff
@@ -0,0 +1,574 @@
+    // then the focus is not contained in it anymore and it should be dismissed
+    // Arguments:
+    // - <unused>
+    // Return Value:
+    // - <none>
+    void ExtensionPalette::_lostFocusHandler(const Windows::Foundation::IInspectable& /*sender*/,
+                                             const Windows::UI::Xaml::RoutedEventArgs& /*args*/)
+    {
+        const auto flyout = _queryBox().ContextFlyout();
+        if (flyout && flyout.IsOpen())
+        {
+            return;
+        }
+
+        auto root = this->XamlRoot();
+        if (!root)
+        {
+            return;
+        }
+
+        auto focusedElementOrAncestor = Input::FocusManager::GetFocusedElement(root).try_as<DependencyObject>();
+        while (focusedElementOrAncestor)
+        {
+            if (focusedElementOrAncestor == *this)
+            {
+                // This palette is the focused element or an ancestor of the focused element. No need to dismiss.
+                return;
+            }
+
+            // Go up to the next ancestor
+            focusedElementOrAncestor = winrt::Windows::UI::Xaml::Media::VisualTreeHelper::GetParent(focusedElementOrAncestor);
+        }
+
+        // We got to the root (the element with no parent) and didn't meet this palette on the path.
+        // It means that it lost the focus and needs to be dismissed.
+        _close();
+    }
+
+    void ExtensionPalette::_previewKeyDownHandler(const IInspectable& /*sender*/,
+                                                  const Windows::UI::Xaml::Input::KeyRoutedEventArgs& e)
+    {
+        const auto key = e.OriginalKey();
+        const auto coreWindow = CoreWindow::GetForCurrentThread();
+        const auto ctrlDown = WI_IsFlagSet(coreWindow.GetKeyState(winrt::Windows::System::VirtualKey::Control), CoreVirtualKeyStates::Down);
+
+        if (key == VirtualKey::Escape)
+        {
+            // Dismiss the palette if the text is empty, otherwise clear the
+            // text box.
+            if (_queryBox().Text().empty())
+            {
+                _close();
+            }
+            else
+            {
+                _queryBox().Text(L"");
+            }
+
+            e.Handled(true);
+        }
+        else if (key == VirtualKey::Enter)
+        {
+            if (const auto& textBox = e.OriginalSource().try_as<TextBox>())
+            {
+                if (!_queryBox().Text().empty())
+                {
+                    _getSuggestions(_queryBox().Text(), _getCurrentLocalTimeHelper());
+                }
+                e.Handled(true);
+                return;
+            }
+            e.Handled(false);
+            return;
+        }
+        else if (key == VirtualKey::C && ctrlDown)
+        {
+            _queryBox().CopySelectionToClipboard();
+            e.Handled(true);
+        }
+        else if (key == VirtualKey::V && ctrlDown)
+        {
+            _queryBox().PasteFromClipboard();
+            e.Handled(true);
+        }
```

Comment:
shouldn't this... just work?

or is this for when the focus isn't in the text box and you want to copy/paste into the text box. okay I think I get it

Reply:
Nope, NOPE. NOPE. TermControl's _Preview Key Handler_ gets ^V before anyone else. Remember, it is awful.